### PR TITLE
OpenCLIP new inference

### PR DESCRIPTION
**Pull request by wanliAlex**
* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

* **What is the current behavior?** (You can also link to an open issue here)

* **What is the new behavior (if this is a feature change)?**

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

* **Have unit tests been run against this PR?** (Has there also been any additional testing?)

* **Related Python client changes** (link commit/PR here)

* **Related documentation changes** (link commit/PR here)

* **Other information**:

* **Please check if the PR fulfills these requirements**
- [ ] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes/features)
- [ ] Docs have been added / updated (for bug fixes / features)

### CODE REVIEW

File: src/marqo/inference/media_download_and_preprocess/media_download_and_preprocess.py [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+    This should be called only if treat URLs as images is True.
+
+    Args:
+        allocated_docs: docs with images to be downloaded by this thread,
+        media_repo: dictionary that will be mutated by this thread. It will add media
+            as values and the URLs as keys
+        tensor_fields: A tuple of tensor_fields. Images will be downloaded for these fields only.
+        media_download_headers: A dict of headers for image download. Can be used
+            to authenticate image downloads
+        force_download: If True, skip the _is_image check and download the fields as images.
```
Comment (papa99do) [inline src/marqo/inference/media_download_and_preprocess/media_download_and_preprocess.py]: there's no force_download arg

File: src/marqo/inference/media_download_and_preprocess/media_download_and_preprocess.py [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+                    preprocessed_image: List[Tensor] = preprocessor.preprocess([image], modality)
+                except OSError as e:
+                    if "image file is truncated" in str(e):
+                        thread_results.append(PreprocessingError(f"Image file is truncated: {url}"))
+                        continue
+                    else:
+                        raise e
+                thread_results.append([(url, preprocessed_image[0])])
+            else:
+                raise ValueError(f"Unexpected image type: {type(image)}")
```
Comment (papa99do) [inline src/marqo/inference/media_download_and_preprocess/media_download_and_preprocess.py]: This causes the whole batch to fail, should we just append this error to the thread_results?

File: src/marqo/inference/native_inference/chunk_download_preprocess_content.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    Returns:
+        Results in the form of a list[list[tuple[str, Tensor]]].
+        The length of the results must be the same as the length of the input content, while each tuple may contain.
+        Each element in the list is a chunk of the content, and each chunk is a list of tuples.
+        Each tuple contains the original content and the preprocessed tensor.
+
+        e.g., The input is ["This is a test sentence", "Test"] with the text prefix "prefix: ", and split by word,
+        The output will be
+            [
+                [("prefix: This is a", tensor), ("prefix: a test sentence", tensor)], # 2 chunks for the first content
```
Comment (papa99do) [inline src/marqo/inference/native_inference/chunk_download_preprocess_content.py]: The chunk should not include the prefix. When used by add_docs, the chunk will need to be saved in Vespa.

File: src/marqo/inference/native_inference/embedding_models/abstract_embedding_model.py [hunk 4]
```diff
@@ -49,6 +45,12 @@ def _check_loaded_components(self):
         pass
 
     @abstractmethod
-    def encode(self):
+    def encode(self, inputs, modality, normalize):
```
Comment (papa99do) [inline src/marqo/inference/native_inference/embedding_models/abstract_embedding_model.py]: nit. Add type hints for params and return type

File: src/marqo/inference/native_inference/encode_content.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+
+def _collect_tensors(preprocessed_content: list[list[tuple[str, Tensor]]]) -> list[Tensor]:
+    collected_tensors = []
+    for chunk in preprocessed_content:
+        if isinstance(chunk, list):
+            for _, tensor in chunk:
+                if isinstance(tensor, Tensor):
+                    collected_tensors.append(tensor)
+        elif isinstance(chunk, (MediaDownloadError, PreprocessingError)):
```
Comment (papa99do) [inline src/marqo/inference/native_inference/encode_content.py:47]: Could it be other types of error?

File: tests/unit_tests/marqo/inference/native_inference/test_chunk_download_preprocess_text_content.py [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+            chunk_config=TextChunkConfig(split_method="word", split_length=3, split_overlap=1),
+            text_prefix="this is a prefix: "
+        )
+        results = _split_prefix_preprocess_text(content, preprocessor, preprocessing_config)
+        self.assertEqual(len(results), 2)
+        result_1 = results[0]
+        self.assertTrue(isinstance(result_1, list))
+        self.assertEqual(len(result_1), 2)
+        self.assertTrue(all(isinstance(t, tuple) for t in result_1))
+        self.assertTrue(all(isinstance(t[0], str) for t in result_1))
```
Comment (papa99do) [inline tests/unit_tests/marqo/inference/native_inference/test_chunk_download_preprocess_text_content.py:51]: should we also test what the chunks are?

File: src/marqo/inference/media_download_and_preprocess/media_download_and_preprocess.py [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+    """
+    _id = f'image_download.{threading.get_ident()}'
+    thread_results: list[Union[MediaDownloadError, PreprocessingError, list[Tuple[str, Tensor]]]] = []
+    with metric_obj.time(f"{_id}.thread_time"):
+        for url in allocated_content:
+            try:
+                image = load_image_from_path(
+                    url, media_download_headers, timeout_ms=download_timeout_ms, metrics_obj=metric_obj
+                )
+            except PIL.UnidentifiedImageError as e:
```
Comment (papa99do) [inline src/marqo/inference/media_download_and_preprocess/media_download_and_preprocess.py:62]: what error is raised when downloading times out?

File: src/marqo/inference/native_inference/inference_pipeline.py [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    def __init__(self, inference_request: InferenceRequest):
+        self.inference_request = inference_request
+
+    def run_pipeline(self) -> InferenceResult:
+        model = load_model(
+            model_name=self.inference_request.model_config.model_name,
+            model_properties=self.inference_request.model_config.model_properties,
+            model_auth=self.inference_request.model_config.model_auth,
+            device=self.inference_request.device
+        )
```
Comment (papa99do) [inline src/marqo/inference/native_inference/inference_pipeline.py]: consider catch all the model related error (which are s2_inference.errors) type, and raise a `inference.api.ModelError` here

File: src/marqo/inference/native_inference/embedding_models/open_clip_model.py [hunk 9]
```diff
@@ -301,7 +301,10 @@ def _download_from_repo(self):
                 f' filepath `{model_file_path}`')
         return model_file_path
 
-    def encode_image(self, images: Tensor, normalize=True) -> np.ndarray:
+    def encode_image(self, images: List[Tensor], normalize=True) -> List[ndarray]:
+
+        images = torch.cat(images, dim=0)
```
Comment (papa99do) [inline src/marqo/inference/native_inference/embedding_models/open_clip_model.py:306]: should we do stacking instead? why do we concatenate images together?

File: src/marqo/inference/native_inference/embedding_models/random_model.py [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+        pass
+
+    def load(self):
+        pass
+
+    def _build_model_properties(self) -> RandomModelProperties:
+        return RandomModelProperties(**self.model_properties)
+
+    def encode(self, inputs: List[str], modality: Modality, normalize: bool = True) -> List[ndarray]:
+        raw_embeddings = np.random.rand(len(inputs), self._model_properties.dimensions)
```
Comment (papa99do) [inline src/marqo/inference/native_inference/embedding_models/random_model.py:42]: do we still need to make sure the same input generate the same embedding here?

File: src/marqo/inference/native_inference/inference_pipeline.py (hunk 8 above)
Comment (wanliAlex) [reply src/marqo/inference/native_inference/inference_pipeline.py]: done

File: src/marqo/inference/native_inference/local_inference.py [hunk 11]
```diff
@@ -1,9 +1,27 @@
...
 class NativeInferenceLocal(Inference):
 
     def vectorise(self, request: InferenceRequest) -> InferenceResult:
-        return InferencePipeline(request).run_pipeline()
+        model = load_model(
+            model_name=request.model_config.model_name,
+            model_properties=request.model_config.model_properties,
+            model_auth=request.model_config.model_auth,
+            device=request.device
+        )
```
Comment (papa99do) [inline src/marqo/inference/native_inference/local_inference.py]: catch model errors from s2_inference.errors, and raise a generic ModelError from inference.api.exceptions?

File: tests/integ_tests/core/inference/test_marqo_fashion_clip.py [hunk 12]
```diff
@@ -1,651 +0,0 @@
...
-import numpy as np
-
-from marqo.inference.native_inference.embedding_models.open_clip_model import OPEN_CLIP
-from marqo.s2_inference.model_registry import _get_open_clip_properties
-from marqo.s2_inference.s2_inference import clear_loaded_models
-
-OPEN_CLIP_MODEL_PROPERTIES = _get_open_clip_properties()
-
-
-class TestMarqoFashionCLIP(TestCase):
```
Comment (papa99do) [inline tests/integ_tests/core/inference/test_marqo_fashion_clip.py:12]: why do we delete it? Can we just skip it?

File: src/marqo/inference/native_inference/chunk_download_preprocess_content.py (hunk 3 above)
Comment (wanliAlex) [reply src/marqo/inference/native_inference/chunk_download_preprocess_content.py]: done

File: src/marqo/inference/native_inference/embedding_models/abstract_embedding_model.py (hunk 4 above)
Comment (wanliAlex) [reply src/marqo/inference/native_inference/embedding_models/abstract_embedding_model.py]: added

File: tests/unit_tests/marqo/inference/native_inference/test_chunk_download_preprocess_text_content.py (hunk 6 above)
Comment (wanliAlex) [reply tests/unit_tests/marqo/inference/native_inference/test_chunk_download_preprocess_text_content.py:51]: done

File: src/marqo/inference/native_inference/embedding_models/open_clip_model.py (hunk 9 above)
Comment (wanliAlex) [reply src/marqo/inference/native_inference/embedding_models/open_clip_model.py:306]: This method converts a list (N) of [1, Dim] tensors to a [N, Dim] tensor, which is the expected input of the model

File: src/marqo/inference/native_inference/embedding_models/random_model.py (hunk 10 above)
Comment (wanliAlex) [reply src/marqo/inference/native_inference/embedding_models/random_model.py:42]: I will discuss this with you offline

File: src/marqo/inference/native_inference/local_inference.py (hunk 11 above)
Comment (wanliAlex) [reply src/marqo/inference/native_inference/local_inference.py]: done

**Review by papa99do — APPROVED** [review]
